File: crates/snolc-abi/tests/header_layout.rs
#![cfg(unix)]

use std::collections::HashMap;
use std::mem::{offset_of, size_of};
use std::path::PathBuf;
use std::process::Command;

use snolc_abi::*;

#[test]
fn c_header_matches_rust_layout() {
    let manifest = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let output = std::env::temp_dir().join(format!("snolc-header-layout-{}", std::process::id()));
    let compiler = std::env::var_os("CC").unwrap_or_else(|| "cc".into());
    let compiled = Command::new(compiler)
        .arg("-std=c11")
        .arg("-Wall")
        .arg("-Wextra")
        .arg("-Werror")
        .arg(manifest.join("tests/header_layout.c"))
        .arg("-I")
        .arg(manifest.join("../../include"))
        .arg("-o")
        .arg(&output)
        .status()
        .expect("C compiler must run");
    assert!(compiled.success());
    let measured = Command::new(&output)
        .output()
        .expect("layout probe must run");
    let _ = std::fs::remove_file(&output);
    assert!(measured.status.success());
    let measured = String::from_utf8(measured.stdout).unwrap();
    let measured: HashMap<_, _> = measured
        .lines()
        .map(|line| {
            let mut fields = line.split_ascii_whitespace();
            let name = fields.next().unwrap();
            let values = fields
                .map(|value| value.parse::<usize>().unwrap())
                .collect::<Vec<_>>();
            (name, values)
        })
        .collect();

    assert_layout::<SnolBytes>(
        &measured,
        "SnolBytes",
        &[
            offset_of!(SnolBytes, pointer),
            offset_of!(SnolBytes, length),
        ],
    );
    assert_layout::<SnolBytesMut>(
        &measured,
        "SnolBytesMut",
        &[
            offset_of!(SnolBytesMut, pointer),
            offset_of!(SnolBytesMut, length),
        ],
    );
    assert_layout::<SnolIoResult>(
        &measured,
        "SnolIoResult",
        &[
            offset_of!(SnolIoResult, tag),
            offset_of!(SnolIoResult, code),
            offset_of!(SnolIoResult, count),
        ],
    );
    assert_layout::<SnolWakeHandle>(
        &measured,
        "SnolWakeHandle",
        &[
            offset_of!(SnolWakeHandle, context),
            offset_of!(SnolWakeHandle, wake),
            offset_of!(SnolWakeHandle, retain),
            offset_of!(SnolWakeHandle, release),
        ],
    );
    assert_layout::<SnolFlowMetadataV1>(
        &measured,
        "SnolFlowMetadataV1",
        &[
            offset_of!(SnolFlowMetadataV1, struct_size),
            offset_of!(SnolFlowMetadataV1, kind),
            offset_of!(SnolFlowMetadataV1, address_type),
            offset_of!(SnolFlowMetadataV1, reserved),
            offset_of!(SnolFlowMetadataV1, address),
            offset_of!(SnolFlowMetadataV1, port),
            offset_of!(SnolFlowMetadataV1, reserved2),
            offset_of!(SnolFlowMetadataV1, metadata),
        ],
    );
    assert_layout::<SnolByteIoV1>(
        &measured,
        "SnolByteIoV1",
        &[
            offset_of!(SnolByteIoV1, struct_size),
            offset_of!(SnolByteIoV1, reserved),
            offset_of!(SnolByteIoV1, read),
            offset_of!(SnolByteIoV1, write),
            offset_of!(SnolByteIoV1, flush),
            offset_of!(SnolByteIoV1, shutdown_write),
            offset_of!(SnolByteIoV1, close),
        ],
    );
    assert_layout::<SnolDatagramIoV1>(
        &measured,
        "SnolDatagramIoV1",
        &[
            offset_of!(SnolDatagramIoV1, struct_size),
            offset_of!(SnolDatagramIoV1, reserved),
            offset_of!(SnolDatagramIoV1, recv_datagram),
            offset_of!(SnolDatagramIoV1, send_datagram),
            offset_of!(SnolDatagramIoV1, close),
        ],
    );
    assert_layout::<SnolAdapterApiV1>(
        &measured,
        "SnolAdapterApiV1",
        &[
            offset_of!(SnolAdapterApiV1, struct_size),
            offset_of!(SnolAdapterApiV1, reserved),
            offset_of!(SnolAdapterApiV1, open),
            offset_of!(SnolAdapterApiV1, accept),
            offset_of!(SnolAdapterApiV1, attach),
            offset_of!(SnolAdapterApiV1, complete),
            offset_of!(SnolAdapterApiV1, close_flow),
            offset_of!(SnolAdapterApiV1, attach_datagram),
            offset_of!(SnolAdapterApiV1, attach_packet_port),
        ],
    );
    assert_layout::<SnolProtectionApiV1>(
        &measured,
        "SnolProtectionApiV1",
        &[
            offset_of!(SnolProtectionApiV1, struct_size),
            offset_of!(SnolProtectionApiV1, reserved),
            offset_of!(SnolProtectionApiV1, wrap),
        ],
    );
    assert_layout::<SnolCarrierApiV1>(
        &measured,
        "SnolCarrierApiV1",
        &[
            offset_of!(SnolCarrierApiV1, struct_size),
            offset_of!(SnolCarrierApiV1, reserved),
            offset_of!(SnolCarrierApiV1, connect),
            offset_of!(SnolCarrierApiV1, accept),
        ],
    );
    assert_layout::<SnolPolicyApiV1>(
        &measured,
        "SnolPolicyApiV1",
        &[
            offset_of!(SnolPolicyApiV1, struct_size),
            offset_of!(SnolPolicyApiV1, reserved),
            offset_of!(SnolPolicyApiV1, attach_session),
            offset_of!(SnolPolicyApiV1, admit_flow),
            offset_of!(SnolPolicyApiV1, attach_flow),
            offset_of!(SnolPolicyApiV1, attach_datagram_flow),
        ],
    );
    assert_layout::<SnolHostApiV1>(
        &measured,
        "SnolHostApiV1",
        &[
            offset_of!(SnolHostApiV1, struct_size),
            offset_of!(SnolHostApiV1, reserved),
            offset_of!(SnolHostApiV1, context),
            offset_of!(SnolHostApiV1, now_monotonic_nanos),
            offset_of!(SnolHostApiV1, set_timer),
            offset_of!(SnolHostApiV1, emit_event),
            offset_of!(SnolHostApiV1, context_get),
            offset_of!(SnolHostApiV1, context_set),
            offset_of!(SnolHostApiV1, protect_socket),
        ],
    );
    assert_layout::<SnolModuleDescriptor>(
        &measured,
        "SnolModuleDescriptor",
        &[
            offset_of!(SnolModuleDescriptor, struct_size),
            offset_of!(SnolModuleDescriptor, wire_version),
            offset_of!(SnolModuleDescriptor, class_mask),
            offset_of!(SnolModuleDescriptor, reserved),
            offset_of!(SnolModuleDescriptor, name),
            offset_of!(SnolModuleDescriptor, describe),
            offset_of!(SnolModuleDescriptor, validate_config),
            offset_of!(SnolModuleDescriptor, create),
            offset_of!(SnolModuleDescriptor, poll),
            offset_of!(SnolModuleDescriptor, control),
            offset_of!(SnolModuleDescriptor, shutdown),
            offset_of!(SnolModuleDescriptor, destroy),
            offset_of!(SnolModuleDescriptor, byte_io),
            offset_of!(SnolModuleDescriptor, datagram_io),
            offset_of!(SnolModuleDescriptor, adapter),
            offset_of!(SnolModuleDescriptor, protection),
            offset_of!(SnolModuleDescriptor, carrier),
            offset_of!(SnolModuleDescriptor, policy),
        ],
    );
}

fn assert_layout<T>(measured: &HashMap<&str, Vec<usize>>, name: &str, offsets: &[usize]) {
    let mut expected = Vec::with_capacity(offsets.len() + 1);
    expected.push(size_of::<T>());
    expected.extend_from_slice(offsets);
    assert_eq!(measured.get(name), Some(&expected), "{name}");
}

File: crates/snolc-abi/tests/header_layout.c
#include <stddef.h>
#include <stdio.h>

#include "snolc.h"

#define FIELD(type, field) printf(" %zu", offsetof(type, field))

int main(void) {
    printf("SnolBytes %zu", sizeof(SnolBytes));
    FIELD(SnolBytes, pointer);
    FIELD(SnolBytes, length);
    putchar('\n');

    printf("SnolBytesMut %zu", sizeof(SnolBytesMut));
    FIELD(SnolBytesMut, pointer);
    FIELD(SnolBytesMut, length);
    putchar('\n');

    printf("SnolIoResult %zu", sizeof(SnolIoResult));
    FIELD(SnolIoResult, tag);
    FIELD(SnolIoResult, code);
    FIELD(SnolIoResult, count);
    putchar('\n');

    printf("SnolWakeHandle %zu", sizeof(SnolWakeHandle));
    FIELD(SnolWakeHandle, context);
    FIELD(SnolWakeHandle, wake);
    FIELD(SnolWakeHandle, retain);
    FIELD(SnolWakeHandle, release);
    putchar('\n');

    printf("SnolFlowMetadataV1 %zu", sizeof(SnolFlowMetadataV1));
    FIELD(SnolFlowMetadataV1, struct_size);
    FIELD(SnolFlowMetadataV1, kind);
    FIELD(SnolFlowMetadataV1, address_type);
    FIELD(SnolFlowMetadataV1, reserved);
    FIELD(SnolFlowMetadataV1, address);
    FIELD(SnolFlowMetadataV1, port);
    FIELD(SnolFlowMetadataV1, reserved2);
    FIELD(SnolFlowMetadataV1, metadata);
    putchar('\n');

    printf("SnolByteIoV1 %zu", sizeof(SnolByteIoV1));
    FIELD(SnolByteIoV1, struct_size);
    FIELD(SnolByteIoV1, reserved);
    FIELD(SnolByteIoV1, read);
    FIELD(SnolByteIoV1, write);
    FIELD(SnolByteIoV1, flush);
    FIELD(SnolByteIoV1, shutdown_write);
    FIELD(SnolByteIoV1, close);
    putchar('\n');

    printf("SnolDatagramIoV1 %zu", sizeof(SnolDatagramIoV1));
    FIELD(SnolDatagramIoV1, struct_size);
    FIELD(SnolDatagramIoV1, reserved);
    FIELD(SnolDatagramIoV1, recv_datagram);
    FIELD(SnolDatagramIoV1, send_datagram);
    FIELD(SnolDatagramIoV1, close);
    putchar('\n');

    printf("SnolAdapterApiV1 %zu", sizeof(SnolAdapterApiV1));
    FIELD(SnolAdapterApiV1, struct_size);
    FIELD(SnolAdapterApiV1, reserved);
    FIELD(SnolAdapterApiV1, open);
    FIELD(SnolAdapterApiV1, accept);
    FIELD(SnolAdapterApiV1, attach);
    FIELD(SnolAdapterApiV1, complete);
    FIELD(SnolAdapterApiV1, close_flow);
    FIELD(SnolAdapterApiV1, attach_datagram);
    FIELD(SnolAdapterApiV1, attach_packet_port);
    FIELD(SnolAdapterApiV1, resolve);
    putchar('\n');

    printf("SnolProtectionApiV1 %zu", sizeof(SnolProtectionApiV1));
    FIELD(SnolProtectionApiV1, struct_size);
    FIELD(SnolProtectionApiV1, flags);
    FIELD(SnolProtectionApiV1, wrap);
    putchar('\n');

    printf("SnolCarrierApiV1 %zu", sizeof(SnolCarrierApiV1));
    FIELD(SnolCarrierApiV1, struct_size);
    FIELD(SnolCarrierApiV1, reserved);
    FIELD(SnolCarrierApiV1, connect);
    FIELD(SnolCarrierApiV1, accept);
    putchar('\n');

    printf("SnolPolicyApiV1 %zu", sizeof(SnolPolicyApiV1));
    FIELD(SnolPolicyApiV1, struct_size);
    FIELD(SnolPolicyApiV1, flags);
    FIELD(SnolPolicyApiV1, attach_session);
    FIELD(SnolPolicyApiV1, admit_flow);
    FIELD(SnolPolicyApiV1, attach_flow);
    FIELD(SnolPolicyApiV1, attach_datagram_flow);
    FIELD(SnolPolicyApiV1, admit_resolved);
    putchar('\n');

    printf("SnolHostApiV1 %zu", sizeof(SnolHostApiV1));
    FIELD(SnolHostApiV1, struct_size);
    FIELD(SnolHostApiV1, reserved);
    FIELD(SnolHostApiV1, context);
    FIELD(SnolHostApiV1, now_monotonic_nanos);
    FIELD(SnolHostApiV1, set_timer);
    FIELD(SnolHostApiV1, emit_event);
    FIELD(SnolHostApiV1, context_get);
    FIELD(SnolHostApiV1, context_set);
    FIELD(SnolHostApiV1, protect_socket);
    putchar('\n');

    printf("SnolModuleDescriptor %zu", sizeof(SnolModuleDescriptor));
    FIELD(SnolModuleDescriptor, struct_size);
    FIELD(SnolModuleDescriptor, wire_version);
    FIELD(SnolModuleDescriptor, class_mask);
    FIELD(SnolModuleDescriptor, reserved);
    FIELD(SnolModuleDescriptor, name);
    FIELD(SnolModuleDescriptor, describe);
    FIELD(SnolModuleDescriptor, validate_config);
    FIELD(SnolModuleDescriptor, create);
    FIELD(SnolModuleDescriptor, poll);
    FIELD(SnolModuleDescriptor, control);
    FIELD(SnolModuleDescriptor, shutdown);
    FIELD(SnolModuleDescriptor, destroy);
    FIELD(SnolModuleDescriptor, byte_io);
    FIELD(SnolModuleDescriptor, datagram_io);
    FIELD(SnolModuleDescriptor, adapter);
    FIELD(SnolModuleDescriptor, protection);
    FIELD(SnolModuleDescriptor, carrier);
    FIELD(SnolModuleDescriptor, policy);
    putchar('\n');
    return 0;
}
